### ISSUE TITLE
Proposal: Increase Account Voting Power Precision

### ISSUE THREAD

**Issue by mvandeberg** [steemit/steem#1808]
The base proposal is to change the voting power field from a `uint16_t` to a `uint32_t` and the base precision from 10,000 to 100,000,000.

We have a rule when calculating VP that at a minimum a vote must consume 0.01% (1 satoshi) VP. However, we have the same precision for VP that we do for vote weight. At 100% VP, a 100% vote used 0.5% VP. So any vote with 2% weight or less, at 100% VP gets treated the same as a 2% vote (rounding up to 1 satoshi). That gets progressively worse as VP is used because votes consume a percentage of VP. The issue is when an account uses all 2% weight votes or less. They are capped at 2000k votes a day.

While this is plenty for normal users, there are third party apps (Cheetah and Chainbb are both running into this problem) that use voting extensively for both internal flags and to affect rewards. This behavior creates a hard limit on the effectiveness of such applications.

We need to increase the precision so that a 0.01% vote at 0.01% VP still has precision.

Given 0.01% VP and a 0.01% vote we have the equation `0.01% * 0.5% * 0.01% = 0.00005%` So we need a precision of 10,000,000 to capture the entire precision of the calculation.

The 0.5% is derived from the `vote_power_reserve_rate` in the dynamic global properties. The effect of the calculation for daily votes is `86400 / ( vote_power_reserve_rate * STEEM_VOTE_REGENERATION_SECONDS )`. There are some assumptions of units in the calculations that makes the code dense that could perhaps be refactored, but would need to be done so carefully to preserve backwards compatibility. Perhaps by adopting the nomenclature of `STEEM_VOTE_REGENERATION_PERIODS = 5` and `STEEM_VOTE_REGENERATION_PERIOD_SECONDS = 86400` so that we could do 10 periods of 12 hours with a reserve rate of 5 to achieve the same results, rather than relying on daily periods.

With this derivation of `0.5%`, we can redefine the precision requirements as `floor( log_10( 1,000,000 * vote_power_reserve_rate * ( STEEM_VOTE_REGENERATION_PERIODS )`. With 5 periods, this means the maximum reserve rate is 20 before needing to increase the precision. Because we have had a reserve rate of 40 in the past and it is reasonable to think that we may want to change it back in the future (this would not be the first reverted change in Steem), we would want a precision of 10^8, which is 100,000,000.

Thankfully, this fits in a 32 bit integer (provides a precision of 10^9). This would increase the size of each account object by 2 bytes. At current size, that is an increase of less than 1 MB.

**Comment by theoreticalbts** [comment]
So basically there is no change to operations, just `account_object.voting_power` is increased to 32 bits and the maximum increased from `10^4` to `10^8`?

Sure, I'd be on board with that.  If we're using a `uint32_t` anyway, why choose `10^8`?  Why not `10^9` or `2^32-1`?  I'd say if we're going to break people who rely on the old scale anyway, let's use the biggest value we can.

**Comment by mvandeberg** [comment]
It is easiest to return the value as a power of 10 (from an API UX perspective). We can increase to 10^9 and have a maximum `vote_power_reserve_rate` of 199.

**Comment by sneak** [comment]
👍

**Comment by E-D-A** [comment]
I was testing out this logic with the minnowsupport bot. We have in the last month reached a number of more than 2000 votes per day and that's how I found out about this reported issue. But the tests I've done doesn't match the description above, I have seen a linear behaviour using vote weights between 0.5% and 2%.

I found in the source code that vote_power_reserve_rate was reduced by 4: 
```
case STEEMIT_HARDFORK_0_19:
         {
            modify( get_dynamic_global_properties(), [&]( dynamic_global_property_object& gpo )
            {
               gpo.vote_power_reserve_rate = 10;
            });
```
It looks like the situation is actually 4 times better. So the current precision wouldn't be enough to differentiate between votes smaller than 0.5% at 100%VP.

I hope I got that right.
